File: example-crs-webservice/crs-userspace/libs/libMSA/python/libmsa/sync/election.py
from kazoo.client import KazooClient


def am_i_leader(zk_hosts: str, leader_path: str) -> bool:
    zk = KazooClient(hosts=zk_hosts)
    zk.start()

    if not zk.exists(leader_path):
        zk.ensure_path(leader_path)

    my_znode = zk.create(f"{leader_path}/node-", sequence=True, ephemeral=False)
    my_znode_name = my_znode.split("/")[-1]

    children = zk.get_children(leader_path)
    children.sort()

    if children[0] == my_znode_name:
        is_leader = True
    else:
        is_leader = False

    zk.stop()

    return is_leader

File: example-crs-webservice/crs-userspace/libs/libMSA/rust/libmsa/src/sync/election.rs
use zookeeper::{Acl, CreateMode, WatchedEvent, Watcher, ZooKeeper};

struct NoopWatcher;

impl Watcher for NoopWatcher {
    fn handle(&self, _event: WatchedEvent) {
        // No operation on events
    }
}

pub fn am_i_leader(zk_hosts: &str, leader_path: &str) -> Result<bool, Box<dyn std::error::Error>> {
    let zk = ZooKeeper::connect(zk_hosts, std::time::Duration::from_secs(15), NoopWatcher)?;

    if zk.exists(leader_path, false)?.is_none() {
        zk.create(leader_path, vec![], Acl::open_unsafe().clone(), CreateMode::Persistent)?;
    }

    let my_znode = zk.create(
        &format!("{}/node-", leader_path),
        vec![],
        Acl::open_unsafe().clone(),
        CreateMode::PersistentSequential,
    )?;

    let my_znode_name = my_znode.split('/').last().unwrap().to_string();

    let mut children = zk.get_children(leader_path, false)?;
    children.sort();

    let is_leader = children.first().map_or(false, |child| child == &my_znode_name);

    zk.close()?;

    Ok(is_leader)
}